pkg/service/model_builder.go

```diff
@@ -214,3 +226,20 @@ func (t *defaultModelBuildTask) getDeletionProtectionViaAnnotation(svc corev1.Se
 	}
 	return false, nil
 }
+
+func (t *defaultModelBuildTask) needsCleanup() bool {
```

Comment:
this is still a little hard to understand this part tbh.

I think maybe like have below method in serviceUtils?
* IsServicePendingFinalization
     -> only checks whether the Service contains our finalizer 
* IsServiceSupported/IsServiceAlive(or other better word than Alive/Supported)
     -> only checks whether the Service is supported, and if so, we'll always provision a NLB.
     -> if !t.service.DeletionTimestamp.IsZero, return false
     -> if config.ServiceTypeLoadBalancerOnly and type not LB, return false
     -> this checks ((spec.IngressClass == service.k8s.aws/nlb) || anno.LBType & anno.NLBTargetType meets requirement)


So that, the event handler will check **whether IsServicePendingFinalization or IsServiceSupported**

And modelBuilder's logic will be like
If !IsServiceSupported{
  if IsServicePendingFinalization {
     checkForDeletionProtection
  }
  return nil
}
return non-nil model

And modelBuilder only need to check whether IsServiceSupported.

Reply:
updated